### DISCUSSION TITLE
Panic after updating fltk-rs

### DISCUSSION THREAD

**Discussion by KillTheMule** [fltk-rs/fltk-rs#1206] [Q&A] [answered]
I've recently updated my fltk-rs dependency from 1.2.16 to 1.3.6, and without changing my own code, I hit the following panic:

> thread 'main' panicked at 'called `Option::unwrap()` on a `None` value', /home/xxx/.cargo/registry/src/github.com-1ecc6299db9ec823/fltk-1.2.22/src/app/init.rs:76:24

As you can see, I've already reduced this to 1.2.22 being the first version this happens on, 1.2.21 is fine. It's not too surprising, since 1.2.22 introduced `is_ui_thread()`.  Note that this happens on both linux and windows.

The problem that on [this line](https://github.com/fltk-rs/fltk-rs/commit/52c5a6a6e12d4c35b524f3664c1de48513baadef#diff-d231fb9347d50a9ce8ed723489e7a0e3b1a941a7edb7b31f47025213b66ca78fR76),  `UI_THREAD` is still `None`, so it has not been initialized.

As far as I can tell, the problem is that my code calls `Window::default()` early on, before I'm actually calling `app::wait()`.

But, isn't that how I'm actually supposed to structure an fltk-rs app? Creating  my windows, then widgets one after another, and when I'm done, I call `window::end()` and then I run the event loop. The counter example works the same way, right? Except that it doesn't panick, and I don't understand why...

What am I doing wrong? Thanks for any pointers :)

**Comment by KillTheMule** [comment, marked as answer]
Ok, just after writing this, I understood what was happening: I was creating the window _before_ even calling `App::default()`, which initializes `UI_THREAD`....

Thanks for reading, I'll investigate if I can cook up a PR that gives a better error message here.

**Reply by KillTheMule** [reply]
Actually, I think this can't happen anymore after switching to `lazy_static`, but it's not yet released and I had trouble using a vendored version, so I'm not sure.

**Reply by MoAlyousef** [reply]
The asserts might be a bit extreme, since even though instantiating a window for example might not work on a non-ui thread, it’s not unsafe per se. I might replace them with a debug_assert, but that will still crash on debug builds, and other Rust gui crates seem to also check whether certain operations are done on the ui thread. 

Regarding vendored versions, you might want to point cargo to the fltk-rs git repo until a new version is released (hopefully next Thursday).

**Reply by KillTheMule** [reply]
Don't sweat anything, as I said, rearranging 2 lines of code solved the problem even without a new release :)
